### PR TITLE
Move database_commom.descendsFrom and prune search using Ref.Height

### PR DESCRIPTION
**Pull request by ehalpern**
- Move to commit.CommitDescendsFrom
- When searching for an ancestor in the commit history, prune
  whenever the commit.Height() < ancestor.Height()
- Add new test TestCommitDescendsFrom to verify correctness.
  Note this does not verify that pruning is occurring.

fixes:#2358

### CODE REVIEW

File: go/datas/commit_test.go [hunk 1]
```diff
@@ -82,3 +84,57 @@ func TestCommitWithoutMetaField(t *testing.T) {
 	})
 	assert.False(IsCommitType(noMetaCommit.Type()))
 }
+
+func TestCommitDescendsFrom(t *testing.T) {
+	assert := assert.New(t)
+	cs := chunks.NewTestStore()
+	db := NewDatabase(cs)
+	defer cs.Close()
```
Comment (aboodman) [inline go/datas/commit_test.go]: Drive-by: do you need to close `cs` manually? I don't usually see this.

File: go/datas/commit_test.go [hunk 2]
```diff
@@ -82,3 +84,57 @@ func TestCommitWithoutMetaField(t *testing.T) {
...
+		commit types.Struct
+		ref    types.Ref
+	}
+	// Add a commit and return commit and ref
+	addCommit := func(ds string, val string, parents types.Set) commitAndRef {
+		commit := NewCommit(types.String(val), parents, types.EmptyStruct)
+		db, err := db.Commit(ds, commit)
+		assert.NoError(err)
+		ref := db.HeadRef(ds)
+		return commitAndRef{commit, ref}
```
Comment (aboodman) [inline go/datas/commit_test.go]: I believe that `commit.Ref() == ref`. So there's no need to store both.

File: go/datas/commit.go [hunk 3]
```diff
@@ -53,6 +53,36 @@ func NewCommit(value types.Value, parents types.Set, meta types.Struct) types.St
...
+	return true
+}
+
+// getAncestors returns set of direct ancestors with height >= minHeight
+func getAncestors(commits types.Set, minHeight uint64, vr types.ValueReader) types.Set {
+	ancestors := types.NewSet()
+	commits.IterAll(func(v types.Value) {
+		r := v.(types.Ref)
+		c := r.TargetValue(vr).(types.Struct)
+		if r.Height() >= minHeight {
```
Comment (rafael-atticlabs) [inline go/datas/commit.go]: i think this is >

File: go/datas/commit.go [hunk 4]
```diff
@@ -53,6 +53,36 @@ func NewCommit(value types.Value, parents types.Set, meta types.Struct) types.St
...
+			next := []types.Value{}
+			c.Get(ParentsField).(types.Set).IterAll(func(v types.Value) {
+				next = append(next, v)
```
Comment (rafael-atticlabs) [inline go/datas/commit.go]: this also needs to check >=

File: go/datas/commit_test.go (hunk 1 above)
Comment (ehalpern) [inline go/datas/commit_test.go]: I copied this pattern from database_test.TearDownTest. If not required, I'll remove from there as well.

File: go/datas/commit_test.go [hunk 5]
```diff
@@ -82,3 +84,57 @@ func TestCommitWithoutMetaField(t *testing.T) {
 	})
 	assert.False(IsCommitType(noMetaCommit.Type()))
 }
+
+func TestCommitDescendsFrom(t *testing.T) {
+	assert := assert.New(t)
```
Comment (rafael-atticlabs) [inline go/datas/commit_test.go]: high-level issue: I think this test would be clearer if it first setup a DAG of commits which collected local variables named `a`, `b`, `c`, etc... (each of these would be a `Ref<Commit>`) and some ascii-art in a comment which depicts the graph, something like: 

https://github.com/attic-labs/noms/blob/master/go/datas/pull_test.go#L168

Then at the bottom a bunch of assertions for the set of cases you need to test, e.g.

`assert.True(descendsFrom(a, b))`
`assert.False(descendsFrom(c, e))`

etc...

**Comment by willhite** [comment]
test this please

File: go/datas/commit_test.go (hunk 5 above)
Comment (ehalpern) [inline go/datas/commit_test.go]: PTAL @rafael-atticlabs. I've restructured the tests in the spirit of your comment.

**Comment by ehalpern** [comment]
@rafael-atticlabs: PYAL to confirm your concerns are addressed. The only change since you reviewed over my shoulder are improvements to the tests.

**Comment by rafael-atticlabs** [comment]
awesome. lgtm. the test is really beautiful. sorry for the delay.

**Comment by ehalpern** [comment]
Thanks for your very helpful feedback.
